File: 04 April/04-08/fizzBuzzValidator.js
function isFizzBuzz(arr) {
  const integersArr = [];
  
  arr.forEach((element, i, source) => {
    if (!isInteger(element)) {
      let integerDetermined;
      
      for (let j = i - 1; j >= 0; --j) {
        if (isInteger(source[j])) {
          integersArr.push(source[j] + i - j);
          integerDetermined = true;
          break;
        }
      }

      if (!integerDetermined) {
        for (let j = i + 1; j < source.length; ++j) {
          if (isInteger(source[j])) {
            integersArr.push(source[j] - (j - i));
            break;
          }
        }
      }
    } else {
      integersArr.push(element);
    }
  })

  for (const i in integersArr) {
    let correctValue = integersArr[i];

    if (!(integersArr[i] % 3) && !(integersArr[i] % 5)) correctValue = "FizzBuzz";
    else if (!(integersArr[i] % 3)) correctValue = "Fizz";
    else if (!(integersArr[i] % 5)) correctValue = "Buzz";

    if (arr[i] !== correctValue) return false;
  }

  return true;
}

const isInteger = element => !Number.isNaN(parseInt(element));

// --- TEST SUITE ---

const testsText = String.raw`
1. isFizzBuzz([1, 2, "Fizz", 4, "Buzz"]) should return true.
2. isFizzBuzz([13, 14, "FizzBuzz", 16, 17]) should return true.
3. isFizzBuzz([1, 2, "Fizz", 4, 5]) should return false.
4. isFizzBuzz(["FizzBuzz", 16, 17, "Fizz", 19, "Buzz"]) should return true.
5. isFizzBuzz([1, 2, "Fizz", "Buzz", 5]) should return false.
6. isFizzBuzz([97, 98, "Buzz", "Fizz", 101, "Fizz", 103]) should return false.
7. isFizzBuzz(["Fizz", "Buzz", 101, "Fizz", 103, 104, "FizzBuzz"]) should return true.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("——————————————————————————",
            "\n🧪Starting Verification...",
            "\n——————————————————————————");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);

    const comparison = Array.isArray(testOutput)
      ? arraysEqual(functionCallOutput, testOutput)
      : functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    } else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      ++failCount;
    }
    console.log("————————————————————————————");
  })

  console.log(failCount
    ? `⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "🎉SUCCESS: All tests PASSED."
  );
}

function arraysEqual(a, b) {
  if (a.length !== b.length) return false;

  for (let i = 0; i < a.length; ++i) {
    if (Array.isArray(a[i])) {
      if (Array.isArray(b[i])) {
        if (!arraysEqual(a[i], b[i])) {
          return false;
        }
      } else {
        return false;
      }
    } else if (a[i] !== b[i]) {
      return false;
    }
  }

  return true;
}

runTests(testData);

File: 04 April/04-08/fizz_buzz_validator.py
def is_fizz_buzz(arr):
    integers_list = []

    for i, element in enumerate(arr):
        if not is_integer(element):
            integer_determined = False

            for j in reversed(range(i)):
                if is_integer(arr[j]):
                    integers_list.append(arr[j] + i - j)
                    integer_determined = True
                    break

            if not integer_determined:
                for j in range(i + 1, len(arr)):
                    if is_integer(arr[j]):
                        integers_list.append(arr[j] - (j - i))
                        break
        else:
            integers_list.append(element)

    for i, integer in enumerate(integers_list):
        correct_value = integer

        if not integer % 3 and not integer % 5:
            correct_value = "FizzBuzz"
        elif not integer % 3:
            correct_value = "Fizz"
        elif not integer % 5:
            correct_value = "Buzz"

        if arr[i] != correct_value:
            return False

    return True

def is_integer(element):
    try:
        int(element)
    except ValueError:
        return False
    else:
        return True

# --- TEST SUITE ---

tests_text = r'''
1. is_fizz_buzz([1, 2, "Fizz", 4, "Buzz"]) should return True.
2. is_fizz_buzz([13, 14, "FizzBuzz", 16, 17]) should return True.
3. is_fizz_buzz([1, 2, "Fizz", 4, 5]) should return False.
4. is_fizz_buzz(["FizzBuzz", 16, 17, "Fizz", 19, "Buzz"]) should return True.
5. is_fizz_buzz([1, 2, "Fizz", "Buzz", 5]) should return False.
6. is_fizz_buzz([97, 98, "Buzz", "Fizz", 101, "Fizz", 103]) should return False.
7. is_fizz_buzz(["Fizz", "Buzz", 101, "Fizz", 103, 104, "FizzBuzz"]) should return True.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("——————————————————————————",
        "\n🧪 Starting Verification...",
        "\n——————————————————————————")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

        print("————————————————————————————")

    print(
        f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)